rtrlib/rtr_mgr.c

```diff
@@ -431,17 +525,156 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
 			    struct spki_record **result,
 			    unsigned int *result_count)
 {
-	return spki_table_get_all(config->groups[0].sockets[0]->spki_table,
+	return spki_table_get_all(config->spki_table,
 				  asn, ski, result, result_count);
 }
 
 void rtr_mgr_stop(struct rtr_mgr_config *config)
 {
+	struct rtr_mgr_group_node *group_node;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
 	MGR_DBG1("rtr_mgr_stop()");
-	for (unsigned int i = 0; i < config->len; i++) {
-		for (unsigned int j = 0; j < config->groups[i].sockets_len; j++)
-			rtr_stop(config->groups[i].sockets[j]);
+	while (node) {
+		group_node = node->data;
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
+		     j++) {
+			rtr_stop(group_node->group->sockets[j]);
+		}
+	node = node->next;
+	}
+	pthread_mutex_unlock(&config->mutex);
+}
+
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+	int err_code = RTR_ERROR;
+
+	pthread_mutex_lock(&config->mutex);
+
+	struct rtr_mgr_group_node *gnode;
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			RTR_DBG("Group with preference value already exists!");
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
+		}
+
+		//TODO This is not pretty. It wants to get the same intervals
+		//that are being used by other groups. Maybe intervals should
+		//be store globally/per-group/per-socket?
+		if (gnode->group->sockets[0]->refresh_interval)
+			refresh_iv = gnode->group->sockets[0]->refresh_interval;
+		if (gnode->group->sockets[0]->retry_interval)
+			retry_iv = gnode->group->sockets[0]->retry_interval;
+		if (gnode->group->sockets[0]->expire_interval)
+			expire_iv = gnode->group->sockets[0]->expire_interval;
+		node = node->next;
```

Comment:
I kind of forgot why we are assigning the intervals in the first place here (even though I think I wrote this).
Wouldn't it be better to leave the intervals at their defaults and add API functionality to set intervals? For now, this might only be for the whole group, e.g.:
`set_refresh_interval(struct rtr_mgr_group *group, unsigned int new_interval)`